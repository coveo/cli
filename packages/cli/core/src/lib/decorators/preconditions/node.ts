import {getBinVersionPrecondition} from './binPreconditionsFactory.js';

export const IsNodeVersionInRange = getBinVersionPrecondition('node', {
  prettyName: 'Node.js',
});
