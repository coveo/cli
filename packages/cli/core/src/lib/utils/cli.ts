import {confirm} from '@coveo/cli-commons/utils/ux';

import globalConfig from '@coveo/cli-commons/config/globalConfig';
import {buildEvent} from '@coveo/cli-commons/analytics/eventUtils';

/**
 *
 * @param message confirmation prompt (yes/no)
 * @param questionName Used for analytics purposes to name the action. Should not contain sensitive information.
 * See https://coveord.atlassian.net/wiki/spaces/RD/pages/2214168207/Amplitude for taxonomy
 * @returns
 */
export async function confirmWithAnalytics(
  message: string,
  questionName: string
): Promise<boolean> {
  const doAction = await confirm(message, false);
  if (doAction) {
    trackEvent(`confirmed ${questionName}`);
    return true;
  } else {
    trackEvent(`cancelled ${questionName}`);
    return false;
  }
}

async function trackEvent(questionName: string) {
  await globalConfig.get().runHook('analytics', {
    event: buildEvent(questionName, {}),
  });
}
