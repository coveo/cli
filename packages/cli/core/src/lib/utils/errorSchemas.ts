import {z} from 'zod';

export const isAggregatedErrorLike = (error: any): error is AggregateError =>
  AggregateErrorLike.safeParse(error).success;

const AggregateErrorLike = z.object({
  name: z.literal('AggregateError'),
  errors: z.array(z.string()),
  message: z.string(),
});

export const isErrorLike = (error: any): error is Error =>
  ErrorLike.safeParse(error).success;

const ErrorLike = z.object({
  name: z.string(),
  message: z.string(),
});
