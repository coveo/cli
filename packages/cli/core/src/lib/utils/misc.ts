import {resolve} from 'path';
import {valid} from 'semver';

export function getPackageVersion(packageName: string) {
  const pathToPackageJson = resolve(
    __dirname,
    ...new Array(3).fill('..'),
    'package.json'
  );
  const pkg = require(pathToPackageJson);
  const dep =
    pkg.dependencies[packageName] ||
    pkg.devDependencies[packageName] ||
    pkg.peerDependencies[packageName];
  const defaultVersion = valid(dep?.toString())!;

  return defaultVersion;
}
