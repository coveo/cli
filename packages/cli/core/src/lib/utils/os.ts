export const appendCmdIfWindows = (cmd: TemplateStringsArray) =>
  `${cmd}${process.platform === 'win32' ? '.cmd' : ''}`;
