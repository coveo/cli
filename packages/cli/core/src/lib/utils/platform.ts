import {Config} from '@coveo/cli-commons/config/config';

export function getTargetOrg(config: Config, target?: string) {
  if (target) {
    return target;
  }
  const cfg = config.get();
  return cfg.organization;
}
