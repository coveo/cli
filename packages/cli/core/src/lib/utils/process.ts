import {spawn, SpawnOptions, ChildProcess} from 'child_process';

function isErrnoException(error: Error): error is NodeJS.ErrnoException {
  return Object.hasOwnProperty.call(error, 'errno');
}

/**
 *
 * @private
 * @param {string} command The command to run
 * @param {string[]} args List of string arguments
 * @param {SpawnOptions} [options={}]
 * @returns {Promise<number>} A promise of the the child process exit code
 */
export function spawnProcess(
  command: string,
  args: string[],
  options: SpawnOptions = {}
): Promise<number> {
  return new Promise((resolve, reject) => {
    spawn(command, args, {
      stdio: ['inherit', 'inherit', 'inherit'],
      ...options,
    }).on('close', (code) => {
      if (code !== 0) {
        // TODO: inject logger decorator so this util method can log
        // this.debug(
        //   `Following command has failed: ${command} ${args.join(' ')}`
        // );

        // Any error message triggered by the execution of the subprocess will be
        // displayed since the terminal is shared by both parent and child processes.
        // So no need to bubble up the error.
        reject(code);
      } else {
        resolve(0);
      }
    });
  });
}

export interface SpawnProcessOutput {
  stdout: string;
  stderr: string;
  exitCode?: string;
}

export async function spawnProcessOutput(
  command: string,
  args: string[],
  options: SpawnOptions = {}
): Promise<SpawnProcessOutput> {
  return new Promise((resolve) => {
    const output: SpawnProcessOutput = {
      stdout: '',
      stderr: '',
    };

    const child = spawn(command, args, options);

    child.stdout?.on('data', (d) => {
      output.stdout += d;
    });

    child.stderr?.on('data', (d) => {
      output.stderr += d;
    });

    const closeListener = (code: number | null) => {
      output.exitCode = Number.isInteger(code) ? String(code) : undefined;
      resolve(output);
    };

    child.on('close', closeListener);

    child.on('error', (e) => {
      output.stderr += e.message;
      if (isErrnoException(e) && e.code) {
        child.off('close', closeListener);
        output.exitCode = e.code;
        resolve(output);
      }
    });
  });
}

export const handleForkedProcess = (subprocess: ChildProcess) => {
  return new Promise<void>((resolve, reject) => {
    subprocess.on('exit', () => {
      resolve();
    });
    subprocess.on('message', (message) => {
      reject(message);
    });
  });
};
