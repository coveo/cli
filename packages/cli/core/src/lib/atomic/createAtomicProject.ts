import {mkdirSync} from 'node:fs';
import {resolve} from 'node:path';
import {Configuration} from '@coveo/cli-commons/config/config';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {platformUrl} from '@coveo/cli-commons/platform/environment';
import {appendCmdIfWindows} from '../utils/os';
import {spawnProcess} from '../utils/process';
import {
  IsAuthenticated,
  AuthenticationType,
  HasNecessaryCoveoPrivileges,
} from '@coveo/cli-commons/preconditions/index';
import {
  createApiKeyPrivilege,
  impersonatePrivilege,
  viewSearchPagesPrivilege,
} from '@coveo/cli-commons/preconditions/platformPrivilege';
import {
  IsNpxInstalled,
  IsNodeVersionInRange,
} from '../decorators/preconditions';
import {getPackageVersion} from '../utils/misc';

interface CreateAppOptions {
  initializerVersion?: string;
  pageId?: string;
  projectName: string;
  cfg: Configuration;
}
export const atomicAppInitializerPackage = '@coveo/create-atomic';
export const atomicLibInitializerPackage =
  '@coveo/create-atomic-component-project';

const supportedNodeVersions = '16.x || 18.x';

const transformPackageNameToNpmInitializer = (packageName: string) =>
  packageName.replace('/create-', '/');

export const atomicLibPreconditions = [
  IsNodeVersionInRange(supportedNodeVersions),
];

export const atomicAppPreconditions = [
  IsAuthenticated([AuthenticationType.OAuth]),
  IsNpxInstalled(),
  IsNodeVersionInRange(supportedNodeVersions),
  HasNecessaryCoveoPrivileges(
    createApiKeyPrivilege,
    impersonatePrivilege,
    viewSearchPagesPrivilege
  ),
];

export async function createAtomicApp(options: CreateAppOptions) {
  const authenticatedClient = new AuthenticatedClient();

  const username = await authenticatedClient.getUsername();
  const cliArgs: string[] = [
    `${atomicAppInitializerPackage}@${
      options.initializerVersion ??
      getPackageVersion(atomicAppInitializerPackage)
    }`,
    '--project',
    options.projectName,
    '--org-id',
    options.cfg.organization,
    '--api-key',
    options.cfg.accessToken!,
    '--platform-url',
    platformUrl({
      environment: options.cfg.environment,
      region: options.cfg.region,
    }),
    '--user',
    username,
  ];

  if (options.pageId) {
    cliArgs.push('--page-id', options.pageId);
  }

  return spawnProcess(appendCmdIfWindows`npx`, cliArgs);
}

interface CreateLibOptions {
  projectName: string;
}

export function createAtomicLib(options: CreateLibOptions) {
  const projectDirectory = resolve(options.projectName);
  mkdirSync(projectDirectory);
  const cliArgs = [
    'init',
    `${transformPackageNameToNpmInitializer(
      atomicLibInitializerPackage
    )}@${getPackageVersion(atomicAppInitializerPackage)}`,
  ];
  return spawnProcess(appendCmdIfWindows`npm`, cliArgs, {
    cwd: projectDirectory,
  });
}
