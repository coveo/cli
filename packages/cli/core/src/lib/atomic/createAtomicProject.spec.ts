jest.mock('node:fs');
import {mkdirSync} from 'node:fs';
jest.mock('node:path');
import {resolve} from 'node:path';

jest.mock('@coveo/cli-commons/platform/authenticatedClient');
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {PlatformEnvironment} from '@coveo/cli-commons/platform/environment';
import {Region} from '@coveo/platform-client';
jest.mock('../utils/process');
import {spawnProcess} from '../utils/process';
jest.mock('../utils/os');
import {appendCmdIfWindows} from '../utils/os';
jest.mock('../utils/misc');
import {getPackageVersion} from '../utils/misc';

import {createAtomicApp, createAtomicLib} from './createAtomicProject';

describe('createAtomicProject', () => {
  const mockedSpawnProcess = jest.mocked(spawnProcess);
  const mockedAppendCmdIfWindows = jest.mocked(appendCmdIfWindows);
  const mockedGetPackageVersion = jest.mocked(getPackageVersion);
  beforeEach(() => {
    jest.resetAllMocks();
    mockedAppendCmdIfWindows.mockImplementation((input) => `${input}`);
    mockedGetPackageVersion.mockReturnValue('1.2.3');
  });

  describe('createAtomicApp()', () => {
    const mockedAuthenticatedClient = jest.mocked(AuthenticatedClient);
    const doMockAuthenticatedClient = () => {
      mockedAuthenticatedClient.mockImplementation(
        () =>
          ({
            createImpersonateApiKey: jest.fn(),
            getUsername: () => Promise.resolve('bob@coveo.com'),
          } as unknown as AuthenticatedClient)
      );
    };

    const callOptions = {
      cfg: {
        accessToken: 'someToken',
        environment: PlatformEnvironment.Dev,
        organization: 'bobbyOrg',
        region: Region.AU,
        version: '1.0.0',
      },
      projectName: 'potato',
    };

    beforeEach(() => {
      doMockAuthenticatedClient();
    });

    describe('without options.pageId', () => {
      it('calls `npx @coveo/create-atomic` properly', async () => {
        await createAtomicApp({...callOptions});
        expect(mockedSpawnProcess).toBeCalledTimes(1);
        expect(mockedSpawnProcess.mock.lastCall).toMatchSnapshot();
      });
    });

    describe('with options.pageId', () => {
      it('calls `npx @coveo/create-atomic` properly', async () => {
        await createAtomicApp({...callOptions, pageId: 'pageId'});
        expect(mockedSpawnProcess).toBeCalledTimes(1);
        expect(mockedSpawnProcess.mock.lastCall).toMatchSnapshot();
      });
    });

    describe('with options.initializerVersion set', () => {
      it('calls `npx @coveo/create-atomic` properly', async () => {
        await createAtomicApp({
          ...callOptions,
          initializerVersion: 'test',
          pageId: 'pageId',
        });
        expect(mockedSpawnProcess).toBeCalledTimes(1);
        expect(mockedSpawnProcess.mock.lastCall).toMatchSnapshot();
      });
    });

    describe('without options.initializerVersion', () => {
      it('calls `npx @coveo/create-atomic` properly and calls getPackageVersion', async () => {
        await createAtomicApp({
          ...callOptions,
          initializerVersion: undefined,
          pageId: 'pageId',
        });
        expect(mockedGetPackageVersion).toBeCalledTimes(1);
        expect(mockedSpawnProcess).toBeCalledTimes(1);
        expect(mockedSpawnProcess.mock.lastCall).toMatchSnapshot();
      });
    });
  });

  describe('createAtomicLib()', () => {
    const mockedResolve = jest.mocked(resolve);
    const mockedMkdirSync = jest.mocked(mkdirSync);
    beforeEach(() => {
      mockedResolve.mockImplementation((input) => input);
    });

    it('should initialize the projectDirectory', async () => {
      await createAtomicLib({projectName: 'kewlProject'});

      expect(mockedResolve).toBeCalledTimes(1);
      expect(mockedResolve).toBeCalledWith('kewlProject');
      expect(mockedMkdirSync).toBeCalledTimes(1);
      expect(mockedMkdirSync).toBeCalledWith('kewlProject');
    });

    it('calls `npx @coveo/create-atomic-component-project` properly', async () => {
      await createAtomicLib({projectName: 'kewlProject'});

      expect(mockedSpawnProcess).toBeCalledTimes(1);
      expect(mockedSpawnProcess.mock.lastCall).toMatchSnapshot();
    });
  });
});
