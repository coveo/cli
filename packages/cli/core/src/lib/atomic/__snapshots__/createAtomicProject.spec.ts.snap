// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAtomicProject createAtomicApp() with options.initializerVersion set calls \`npx @coveo/create-atomic\` properly 1`] = `
[
  "npx",
  [
    "@coveo/create-atomic@test",
    "--project",
    "potato",
    "--org-id",
    "bobbyOrg",
    "--api-key",
    "someToken",
    "--platform-url",
    "https://platformdev-au.cloud.coveo.com",
    "--platform-environment",
    "dev",
    "--user",
    "bob@coveo.com",
    "--page-id",
    "pageId",
  ],
]
`;

exports[`createAtomicProject createAtomicApp() with options.pageId calls \`npx @coveo/create-atomic\` properly 1`] = `
[
  "npx",
  [
    "@coveo/create-atomic@1.2.3",
    "--project",
    "potato",
    "--org-id",
    "bobbyOrg",
    "--api-key",
    "someToken",
    "--platform-url",
    "https://platformdev-au.cloud.coveo.com",
    "--platform-environment",
    "dev",
    "--user",
    "bob@coveo.com",
    "--page-id",
    "pageId",
  ],
]
`;

exports[`createAtomicProject createAtomicApp() without options.initializerVersion calls \`npx @coveo/create-atomic\` properly and calls getPackageVersion 1`] = `
[
  "npx",
  [
    "@coveo/create-atomic@1.2.3",
    "--project",
    "potato",
    "--org-id",
    "bobbyOrg",
    "--api-key",
    "someToken",
    "--platform-url",
    "https://platformdev-au.cloud.coveo.com",
    "--platform-environment",
    "dev",
    "--user",
    "bob@coveo.com",
    "--page-id",
    "pageId",
  ],
]
`;

exports[`createAtomicProject createAtomicApp() without options.pageId calls \`npx @coveo/create-atomic\` properly 1`] = `
[
  "npx",
  [
    "@coveo/create-atomic@1.2.3",
    "--project",
    "potato",
    "--org-id",
    "bobbyOrg",
    "--api-key",
    "someToken",
    "--platform-url",
    "https://platformdev-au.cloud.coveo.com",
    "--platform-environment",
    "dev",
    "--user",
    "bob@coveo.com",
  ],
]
`;

exports[`createAtomicProject createAtomicLib() calls \`npx @coveo/create-atomic-component-project\` properly 1`] = `
[
  "npm",
  [
    "init",
    "@coveo/atomic-component-project@1.2.3",
  ],
  {
    "cwd": "kewlProject",
  },
]
`;
